### ISSUE TITLE
loops are not counted in `optLoopCount` in optUnrollLoops phase.

### ISSUE THREAD

**Issue by ArtBlnd** [dotnet/coreclr#19617]
I was working on partial unrolling on master branch. I tried my own test-cases that was working on old branch which I was working on old PR. on last branch almost all of common loops are not counted. and it means it does not optimized out. 

Is there are any changes on optLoopCount? Thanks :>

**Comment by mikedn** [comment]
Hmm, I'm not sure I understand. Are you saying that something changed in the JIT since you started working on unroll and now counted loops are more rare?

I'm not aware of any recent changes in the loop area. Though I suppose there could have been other changes that caused this and passed unnoticed due to the fact that the JIT doesn't do a lot with loops anyway.

**Comment by ArtBlnd** [comment]
Yes, Its lot more rare than older diffs, but those rare ones are really complicate loops.
that causes these all coreclr tests to not unrolled.

also. all my tests are not unrolled. doesn't matter with or without my PR.

**Comment by ArtBlnd** [comment]
lol, okay that only happens when its debug build.

**Comment by mikedn** [comment]
> that only happens when its debug build.

Are you looking at code in System.Private.CoreLib? Then yeah, it's likely. In debug build it has optimizations disabled.

**Comment by ArtBlnd** [comment]
no, I mean `optLoopCount`.

**Comment by ArtBlnd** [comment]
`optLoopCount` isn't counted when CoreCLR is on debug build. not a target sample.

**Comment by mikedn** [comment]
That seems unlikely. A debug build of the JIT is a normal build with a bunch of asserts and other debug checks enabled, it's not a build with JIT optimizations phases disabled or removed. It wouldn't make any sense to have a debug build that somehow doesn't behave like a release build.

**Comment by ArtBlnd** [comment]
![image](https://user-images.githubusercontent.com/2209400/44594468-3f958f80-a801-11e8-8b82-425a8e30b293.png)
This is debug build. As you can see optLoopCount is zero.

![image](https://user-images.githubusercontent.com/2209400/44594547-78cdff80-a801-11e8-950b-d33af3ea758e.png)
This is checked build. its counted as 2. and its a same test.

JIT/Directed/UnrollLoop/loop4_cs_ro

**Comment by ArtBlnd** [comment]
![image](https://user-images.githubusercontent.com/2209400/44594689-e67a2b80-a801-11e8-844d-6d5fb584ca79.png)
debug build

![image](https://user-images.githubusercontent.com/2209400/44594712-f85bce80-a801-11e8-9298-9408aba95d53.png)
checked build

I know it doesn't make any sense, but as you can see. :< Feels bad.

**Comment by AndyAyersMS** [comment]
@ArtBlnd I see the same behavior with debug and checked, at least on the loop4_cs_ro case. And this test case is so large the jit decides not to optimize it at all, so it's no longer a good test for loop unrolling.
```
CLFLG_MINOPT set for method SmallLoop1:Main():int
IL Code Size,Instr 2654,1322, Basic Block count 243, Local Variable Num,Ref count  51,482
for method SmallLoop1:Main():int
```
(Tried debug/checked builds of tests with debug/checked builds of core -- all 4 variations agree).

Are you only seeing this behavior for jits where you've made changes? If so perhaps at some point you need to check whether the jit is actually optimizing.

We should probably break up this test case into smaller methods so it actually does test unrolling. That might be a good project for you, since we will want to have good test cases....

**Comment by ArtBlnd** [comment]
Um, In fact, it appears where I fixed the code.
but I don't touch any of `optLoopCount` variable, also it is counted as zero before calling `optUnrollLoops` function's body.

**Comment by ArtBlnd** [comment]
also, nothing recorded over `optLoopTable`

**Comment by AndyAyersMS** [comment]
Are you referring to your changes over in #19594? I looked those over and also didn't spot anything that looked like it might cause this behavior.

Can you compare the jit dumps for the checked and debug cases like I did? That might help pinpoint what is going on.

**Comment by ArtBlnd** [comment]
I've tried to debug it for a while today. and it seems those IRs are not resolved as loop.
so it means that is not problem of optimization phase. 
even flowgraph on compiler doesn't have loop descriptions.

> Are you referring to your changes over in #19594? I looked those over and also didn't spot anything that looked like it might cause this behavior.

Can you compare the jit dumps for the checked and debug cases like I did? That might help pinpoint what is going on.

I think this problem needs full dump logs. imma dump it ASAP. thanks :>

**Comment by ArtBlnd** [comment]
https://1drv.ms/t/s!Avjm8UvmA5M9gdIdI5GyX-OCuaKB2w
Here is full dump of loop4_cs_ro testcase with debug build

the problem seems these BasicBlocks are not marked(flagged) as loop.

**Comment by ArtBlnd** [comment]
But on this code. still opts.MinOps() is enabled. even its not a small code.

```
using System.Runtime.CompilerServices;

namespace OptLoopArr
{
    class Program
    {
        static int Main(string[] args)
        {
            int[] Arr = new int[256];

            for (int i = 0; i < 256; ++i)
            {
                Arr[i] = Arr[i] + 1;
            }

            return Test(Arr);
        }

        [MethodImpl(MethodImplOptions.NoInlining)]
        static unsafe int Test(int[] arr)
        {
            int total = 0;

            fixed (int* pArr = &arr[0])
            {
                for (int i = 0; i < 256; ++i)
                {
                    total += pArr[i];
                }
            }

            return total;
        }
    }
}
```

![image](https://user-images.githubusercontent.com/2209400/44636987-9231a000-a9e9-11e8-8d1a-ea9dec5f73cd.png)

**Comment by ArtBlnd** [comment]
seems `JIT_FLAG_TIER0`(`CORJIT_FLAG_TIER0`) flag causes this problem.
not because of large code detection.

**Comment by AndyAyersMS** [comment]
Because we've enabled tiered compilation by default in master, you will now need to set `COMPlus_TieredCompilation=0` to get methods to optimize the first time they're called.

**Comment by ArtBlnd** [comment]
Does it really make sense that tiered compilation does only in debug build?
just for sure :>

and btw, problem has been fixed.
